Create gridpack::environment with existing MPI communicator

wperkins opened the pull request:
This branch adds a facility to create a `gridpack::environment` with an existing MPI communicator.  This allows an external application to call `MPI_Initialize()` and instantiate multiple GridPACK environments on an externally distribution of MPI ranks.  
**NOTE**: This GridPACK branch depends on a specific branch of [Global Arrays](https://github.com/GlobalArrays/ga/tree/develop) (`feature/init-comm`).  This PR should not be accepted until the necessary changes have been merged into the GA `master` branch (and the GridPACK GA submodule is updated and actually works).

Commit:
Added a constructor for the environment that takes a communicator for an
argument and an example code.

Commit:
Added a constructor for the environment that takes a communicator for an
argument and an example code.

Commit:
Add example for communicator-based environment

Commit:
Added a configuration option so that branch still builds if a GA library that
does not support initialization with a communicator is used.

Commit:
Added a configuration option so that branch still builds if a GA library that
does not support initialization with a communicator is used.

Commit:
Fixed up some bugs with environment constructor that takes a communicator as an
argument.

Commit:
Fixed up some issues if not compiling with communicator based constructor for
environment.

wperkins committed:
Minor changes to find GA

wperkins committed:
Add a split communicator test of the new Environment constructor

wperkins committed:
Add Python interface to new Environment constructor

Commit:
Fixed a bug when instantiating the environment with a communicator.

wperkins committed:
Add test hadrec script that uses external communicator

wperkins committed:
Change GA url and point to required feature/init-comm branch

wperkins committed:
Fix typo so option works correctly

wperkins committed:
Run init-comm environment test if called for

wperkins committed:
Always build with environment with comm

wperkins committed:
GridPACK needs to know GA headers are installed in .../include/ga

wperkins committed:
Change install directory for my system

wperkins committed:
Make sure mpi4py is available before setup

wperkins committed:
No, don't make sure mpi4py is available for setup, do it elsewhere

wperkins committed:
Check Python and mpi4py availability separate from pybind11

wperkins committed:
Switch GA submodule to develop

abhyshr commented:
1. This branch needs to resolve the conflicts first.
2. Don't understand what is the additional functionality being added here. Is the ability to set an MPI_Comm with GridPACK's environment object & being able to set it via Python?

wperkins commented:
> 2\. Don't understand what is the additional functionality being added here. Is the ability to set an MPI_Comm with GridPACK's environment object & being able to set it via Python?

GridPACK expects to control MPI, that is GridPACK calls `MPI_Init()` and `MPI_Finalize()` to set up an `environment`.  The changes in this PR simply allows a GridPACK `environment` to be created with an externally created `MPI_Comm`.  And, not just in Python. 

This was set up for the [ExaLearn](https://github.com/exalearn) folks, who drive multiple GridPACK instances with an MPI-based  driver.

wperkins commented:
> 1. This branch needs to resolve the conflicts first.

Agreed.  It should be noted, however, that this will require reconciling the HADREC application with current dynamic simulation API without alienating the [ExaLearn](https://github.com/exalearn) application of HADREC.  I do not know details.

wperkins commented:
Note to self: The value of the CMake variable `ENABLE_ENVIRONMENT_FROM_COMM` needs to be in `GridPACK.cmake` and the Python interface adjusted to allow `ENABLE_ENVIRONMENT_FROM_COMM=FALSE`.

Commit:
Modified environment so that no preprocessor symbols are in header. These may
confuse builds outside of gridpack.

wperkins commented on src/environment/environment.cpp:
```diff
@@ -83,22 +86,65 @@ Environment::Environment(int argc, char **argv,const char* help,const long int&
 {
-  // Finalize math libraries
-  gridpack::math::Finalize();
+  if (p_from_comm) {
```
Note the artifacts from rebase.
I don't understand the destructor.  It seems to me that `gridpack::math::Finalize();` and `GA_Terminate()` should be called no matter the state of `p_from_comm`.

wperkins commented on src/applications/dynamic_simulation_full_y/CMakeLists.txt:
```diff
@@ -198,4 +215,11 @@ install(TARGETS dsf2.x DESTINATION bin)
 # run application as test
 # -------------------------------------------------------------
 gridpack_add_run_test("dynamic_simulation_full_y" dsf.x input_145.xml)
+<<<<<<< HEAD
```
Rebase artifacts.

bjpalmer commented:
I'm in the process of cleaning up after the rebase. I'm not sure how I managed to overlook some of the artifacts since the code doesn't compile. I noticed the same thing about the destructor and fixed it.

wperkins commented on src/applications/dynamic_simulation_full_y/dsf_comm_main.cpp:
```diff
@@ -0,0 +1,174 @@
+    //gridpack::utility::Configuration::CursorPtr cursor;
+    cursor = config->getCursor("Configuration.Dynamic_simulation");
+    std::vector<gridpack::dynamic_simulation::Event> faults;
+    faults = ds_app.getFaults(cursor);
```
Also, `getFaults()` is no longer a method.  I think it's `getEvents()` now. I don't know if there is an internal difference.

bjpalmer replied:
FIxed.

Commit:
Fixed up a bunch of problems after rebase, removed some debugging prints from
some of the code and cleaned up the environment destructor so that it works
properly for regular environments.

wperkins committed:
Only warn about GA when using an external build

wperkins committed:
Boost and MPI flags are necessary, don't comment them out

Commit:
Tweaked a few things so that environment can exit cleanly from applications that
are initialized by a communicator and to prevent problems with tests using the
GA progress ranks runtime.

abhyshr committed:
Updated the 39-bus python example to run in parallel. The application hangs/gets stuck

abhyshr commented:
@bjpalmer: any updates on when GA will be updated to have fewture/init-comm in develop?

bjpalmer commented:
The stuff in feature/init-comm should already be in the develop branch of GA. I've been linking GridPACK against the develop branch and it seems to work okay.

wperkins commented:
> The stuff in feature/init-comm should already be in the develop branch of GA. I've been linking GridPACK against the develop branch and it seems to work okay.

FYI. In the GridPACK `feature/init-comm` branch, if you specify `-D BUILD_GA=YES` the GA `develop` branch will be built.

abhyshr commented:
So then can we just use the develop branch of GA? And merge GridPack's feature/init-comm to develop? I can update the install script to use GA's develop branch. Thoughts?

abhyshr commented:
First we need to resolve the issue with GridPACK python application hanging on multiple processors. @bjpalmer: do you have time to take a look at this?

bjpalmer commented:
You should be able to switch the install script to use develop. Is the code hanging when using the progress ranks runtime in GA? You need to be careful with that runtime to get the code to exit cleanly.

abhyshr commented:
I am not using progress ranks. You can try the updated 39 bus python example I've pushed to the branch

wperkins committed:
Set installation directory from environment

wperkins commented:
> You should be able to switch the install script to use develop. Is the code hanging when using the progress ranks runtime in GA? You need to be careful with that runtime to get the code to exit cleanly.

You can still use the official GA release to build this branch. But, if you do, you cannot specify `-D ENABLE_ENVIRONMENT_FROM_COMM=TRUE`. It would be nice to rework `FindGA.cmake` to test for the necessary GA routine and set that option automatically. That's a lot of work though. 

Personally, I think we should merge this. The audience for `-D ENABLE_ENVIRONMENT_FROM_COMM=TRUE` is small and they are already navigating it.

wperkins committed:
Tweak Python installation

wperkins committed:
Add some brief documentation for `ENABLE_ENVIRONMENT_FROM_COMM`

wperkins commented:
@bjpalmer, I've added some very brief documentation.  If it's enough, please approve this PR and I'll merge it.  Thanks.

abhyshr commented:
Good to merge after rebase.
